### PR TITLE
Farm contract with traits

### PR DESCRIPTION
**Pull request by dorin-iancu**

### CODE REVIEW

File: common/modules/farm/events/src/events.rs [hunk 1]
```diff
@@ -3,12 +3,12 @@
...
 elrond_wasm::derive_imports!();
 
-use common_structs::{DefaultFarmPaymentAttributesPair, FarmTokenAttributes};
-use contexts::{
-    claim_rewards_context::{ClaimRewardsContext, CompoundRewardsContext},
-    exit_farm_context::ExitFarmContext,
-    storage_cache::{FarmContracTraitBounds, StorageCache},
-};
+use common_structs::FarmTokenAttributes;
+// use contexts::{
```
Comment (sasurobert) [inline common/modules/farm/events/src/events.rs:7]: why is everything commented out ?

File: common/modules/farm/farm_base_impl/src/base_traits_impl.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + Mergeable<<Self::FarmSc as ContractBase>::Api>
+        + FixedSupplyToken<<Self::FarmSc as ContractBase>::Api>
+        + FarmToken<<Self::FarmSc as ContractBase>::Api> =
+        FarmTokenAttributes<<Self::FarmSc as ContractBase>::Api>;
+
+    #[inline]
+    fn mint_rewards(
```
Comment (sasurobert) [inline common/modules/farm/farm_base_impl/src/base_traits_impl.rs:35]: we actually do not have any mint here - we do not want mint anymore for new contracts as they give out LKMEX created by simple lock energy.

File: common/modules/farm/farm_base_impl/src/exit_farm.rs [hunk 3]
```diff
@@ -26,67 +25,63 @@ pub trait BaseExitFarmModule:
...
+        let exit_farm_context = ExitFarmContext::<Self::Api, FC::AttributesType>::new(
+            payment,
+            &storage_cache.farm_token_id,
+            self.blockchain(),
+        );
 
         self.validate_contract_state(storage_cache.contract_state, &storage_cache.farm_token_id);
-        generate_rewards_fn(self, &mut storage_cache);
+        FC::generate_aggregated_rewards(self, &mut storage_cache);
+        // wrapper.sc_ref.generate_aggregated_rewards(&mut storage_cache)
```
Comment (sasurobert) [inline common/modules/farm/farm_base_impl/src/exit_farm.rs:49]: delete this.

File: common/modules/farm/rewards/src/rewards.rs [hunk 4]
```diff
@@ -38,7 +38,7 @@ pub trait RewardsModule:
         if current_block_nonce > last_reward_nonce {
             let to_mint = self.calculate_per_block_rewards(current_block_nonce, last_reward_nonce);
             if to_mint != 0 {
-                mint_function(token_id, &to_mint);
+                mint_function(self, token_id, &to_mint);
```
Comment (sasurobert) [inline common/modules/farm/rewards/src/rewards.rs:41]: we need to change - as no more mint here.

File: common/modules/farm/farm_base_impl/src/base_traits_impl.rs (hunk 2 above)
Comment (psorinionut) [reply common/modules/farm/farm_base_impl/src/base_traits_impl.rs:35]: Maybe we update this in a future PR, one that includes the entire flow, with both farm & simple-lock-energy contract updates.

Comment (sasurobert) [reply common/modules/farm/farm_base_impl/src/base_traits_impl.rs:35]: I think we should first focus on the FARM we need, that one does not make mint and ask simple lock with energy contract to create and send tokens to the user. Let's do that first. Later we can come back and make something else.

File: common/modules/farm/events/src/events.rs (hunk 1 above)
Comment (dorin-iancu) [reply common/modules/farm/events/src/events.rs:7]: Events were unused, and the compiler kept giving me warnings. Will integrate them again in a future PR.

File: common/modules/farm/farm_base_impl/src/base_traits_impl.rs (hunk 2 above)
Comment (dorin-iancu) [reply common/modules/farm/farm_base_impl/src/base_traits_impl.rs:34]: Mint function can be easily overwritten with an empty function, so no actual token minting is done. Will do this when the optimization for "no needless mint/burn" PR.

File: common/modules/farm/rewards/src/rewards.rs (hunk 4 above)
Comment (dorin-iancu) [reply common/modules/farm/rewards/src/rewards.rs:41]: Same as above.

File: dex/farm/src/base_functions.rs [hunk 5]
```diff
@@ -200,27 +187,24 @@ pub trait BaseFunctionsModule:
...
+            farm_token_amount * &rps_diff / &storage_cache.division_safety_constant
+        } else {
+            BigUint::zero()
+        };
+
         let boosted_yield_rewards =
-            self.claim_boosted_yields_rewards(user, &storage_cache.reward_token_id);
-        base_farm_reward + boosted_yield_rewards
-    }
+            sc.claim_boosted_yields_rewards(&caller, &storage_cache.reward_token_id);
```
[the remark this answers is not in the document]
Comment (dorin-iancu) [reply dex/farm/src/base_functions.rs:198]: That's exactly how it works :)

**Review by psorinionut — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

File: common/modules/farm/farm_base_impl/src/exit_farm.rs [hunk 6]
```diff
@@ -26,67 +25,62 @@ pub trait BaseExitFarmModule:
...
-        self.burn_farm_token_payment(&exit_farm_context.farm_token.payment);
+        let mut reward = FC::calculate_rewards(
+            self,
+            caller,
+            farm_token_amount,
+            &token_attributes,
+            &storage_cache,
+        );
+        storage_cache.reward_reserve -= &reward;
+        reward += token_attributes.get_compounded_rewards();
```
Comment (claudiu725) [inline common/modules/farm/farm_base_impl/src/exit_farm.rs:65]: Shouldn't the reward reserve be updated after the final reward amount, which includes the compounded rewards, is computed?

File: common/traits/mergeable/src/lib.rs [hunk 7]
```diff
@@ -62,3 +62,13 @@ impl<M: ManagedTypeApi> ExternallyMergeable<M> for EsdtTokenPayment<M> {
...
+
+pub fn weighted_average<M: ManagedTypeApi>(
+    first_value: &BigUint<M>,
+    first_weigth: &BigUint<M>,
+    second_value: &BigUint<M>,
+    second_weigth: &BigUint<M>,
+) -> BigUint<M> {
+    let weigth_sum = first_weigth + second_weigth;
+    (first_value * first_weigth + second_value * second_weigth) / weigth_sum
+}
```
Comment (claudiu725) [inline common/traits/mergeable/src/lib.rs:74]:
```suggestion
pub fn weighted_average<M: ManagedTypeApi>(
    first_value: &BigUint<M>,
    first_weight: &BigUint<M>,
    second_value: &BigUint<M>,
    second_weight: &BigUint<M>,
) -> BigUint<M> {
    let weight_sum = first_weight + second_weight;
    (first_value * first_weight + second_value * second_weight) / weight_sum
}
```
Typo fix: weigth -> weight

File: dex/farm/src/base_functions.rs [hunk 8]
```diff
@@ -153,44 +104,72 @@ pub trait BaseFunctionsModule:
...
-        self.burn_farm_tokens_from_payments(&payments);
+        let token_mapper = self.farm_token();
+        let mut output_attributes: FarmTokenAttributes<Self::Api> = self.get_attributes_as_part_of_fixed_supply(&first_payment, &token_mapper);
+        token_mapper.nft_burn(first_payment.token_nonce, &first_payment.amount);
 
-        self.mint_farm_tokens(farm_token_id, attrs.current_farm_amount.clone(), &attrs)
+        for payment in &payments {
+            let attributes: FarmTokenAttributes<Self::Api> = self.get_attributes_as_part_of_fixed_supply(&payment, &token_mapper);
+            output_attributes.merge_with(attributes);
+        }
```
Comment (claudiu725) [inline dex/farm/src/base_functions.rs:117]: Refactor suggestion as to eliminate some code duplication - group these into an utility function:
`self.merge_payments(&mut output_attributes, &payments);`

**Review by claudiu725 — CHANGES_REQUESTED** [review]

File: common/modules/farm/farm_base_impl/src/exit_farm.rs (hunk 6 above)
Comment (dorin-iancu) [reply common/modules/farm/farm_base_impl/src/exit_farm.rs:65]: Nope, during compound_rewards, the reward_reserve is already decremented before being added to the compounded amount.

File: dex/farm/src/base_functions.rs (hunk 8 above)
Comment (dorin-iancu) [reply dex/farm/src/base_functions.rs:117]: Done.

**Review by sasurobert — DISMISSED** [review]

**Review by claudiu725 — APPROVED** [review]

**Review by psorinionut — APPROVED** [review]

**Review by CostinCarabas — APPROVED** [review]

**Pull request merged**